fe/fe-core/src/main/java/com/starrocks/transaction/PublishVersionDaemon.java

```diff
@@ -334,6 +385,210 @@ private CompletableFuture<Void> publishLakeTransactionAsync(TransactionState txn
+                return true;
+            }
+            if (partition.getVisibleVersion() + 1 != versions.get(0)) {
+                LOG.info("publish partition batch partition.getVisibleVersion() + 1 != version.get(0)" + " "
+                        + partition.getId() + " " + partition.getVisibleVersion() + " " + versions.get(0));
+                return false;
+            }
+
+            Set<Tablet> normalTablets = null;
+            Set<Tablet> shadowTablets = null;
+
+            for (int i = 0; i < transactionStates.size(); i++) {
+                TransactionState txnState = transactionStates.get(i);
+                List<MaterializedIndex> indexes = txnState.getPartitionLoadedTblIndexes(table.getId(), partition);
+                for (MaterializedIndex index : indexes) {
+                    if (!index.visibleForTransaction(txnState.getTransactionId())) {
+                        LOG.info("Ignored index {} for transaction {}", table.getIndexNameById(index.getId()),
+                                txnState.getTransactionId());
+                        continue;
+                    }
+                    if (index.getState() == MaterializedIndex.IndexState.SHADOW) {
+                        shadowTablets = (shadowTablets == null) ? Sets.newHashSet() : shadowTablets;
+                        shadowTablets.addAll(index.getTablets());
+                        if (shadowTabletsMap.containsKey(versions.get(i))) {
+                            shadowTabletsMap.get(versions.get(i)).addAll(index.getTablets());
+                        } else {
+                            Set<Tablet> tabletsNew = new HashSet<>(index.getTablets());
+                            shadowTabletsMap.put(versions.get(i), tabletsNew);
+                        }
+
+                    } else {
+                        normalTablets = (normalTablets == null) ? Sets.newHashSet() : normalTablets;
+                        normalTablets.addAll(index.getTablets());
+                    }
+                }
+
+            }
+
+            long startVersion = versions.get(0);
+            long endVersion = versions.get(versions.size() - 1);
+
+            try {
+                for (Map.Entry<Long, Set<Tablet>> item : shadowTabletsMap.entrySet()) {
+                    int index = txnIds.indexOf(item.getKey());
+                    List<Tablet> publishShdowTablets = new ArrayList<>();
+                    publishShdowTablets.addAll(normalTablets);
+                    Utils.publishLogVersionBatch(publishShdowTablets, txnIds.subList(index, txnIds.size()),
+                            versions.subList(index, versions.size()));
+                }
+                if (CollectionUtils.isNotEmpty(normalTablets)) {
+                    Map<Long, Double> compactionScores = new HashMap<>();
+                    List<Tablet> publishTablets = new ArrayList<>();
+                    publishTablets.addAll(normalTablets);
+                    Utils.publishVersion(publishTablets, txnIds,
```

Comment:
why not call publishVersionBatch()?

Reply:
For original publishVersion already support multi txns, i will rename the functionName